ops/testing.py

```diff
@@ -834,5 +857,332 @@ def network_get(self, endpoint_name, relation_id=None):
+                if service.override not in ('merge', 'replace'):
+                    raise RuntimeError('500 Internal Server Error: layer "{}" has invalid '
+                                       '"override" value on service "{}"'.format(label, name))
+                if service.override != 'replace':
+                    raise RuntimeError(
+                        'override: "{}" unsupported for layer "{}" service "{}"'.format(
+                            service.override, label, name))
+                layer.services[name] = service
+        else:
+            self._layers[label] = layer_obj
+
+    def _render_services(self) -> typing.Mapping[str, pebble.Service]:
+        services = {}
+        for key in sorted(self._layers.keys()):
+            layer = self._layers[key]
+            for name, service in layer.services.items():
+                # TODO: (jam) 2021-04-07 have a way to merge existing services
+                services[name] = service
+        return services
+
+    def get_plan(self) -> pebble.Plan:
+        """Get the Pebble plan (currently contains only combined services)."""
+        plan = pebble.Plan('{}')
+        services = self._render_services()
+        if not services:
+            return plan
+        for name in sorted(services.keys()):
+            plan.services[name] = services[name]
+        return plan
+
+    def get_services(self, names: typing.List[str] = None) -> typing.List[pebble.ServiceInfo]:
+        """Get the service status for the configured services.
+
+        If names is specified, only fetch the service status for the services
+        named.
+        """
+        if isinstance(names, str):
+            raise TypeError('start_services should take a list of names, not just "{}"'.format(
+                names))
+        services = self._render_services()
+        infos = []
+        if names is None:
+            names = sorted(services.keys())
+        for name in sorted(names):
+            try:
+                service = services[name]
+            except KeyError:
+                # in pebble, it just returns "nothing matched" if there are 0 matches,
+                # but it ignores services it doesn't recognize
+                continue
+            status = self._service_status.get(name, pebble.ServiceStatus.INACTIVE)
+            if service.startup == '':
+                startup = pebble.ServiceStartup.DISABLED
+            else:
+                startup = pebble.ServiceStartup(service.startup)
+            info = pebble.ServiceInfo(name,
+                                      startup=startup,
+                                      current=pebble.ServiceStatus(status))
+            infos.append(info)
+        return infos
+
+    def pull(self, path: str, *, encoding: str = 'utf-8') -> typing.Union[typing.BinaryIO,
+                                                                          typing.TextIO]:
+        """Read a file's content from the remote system.
```

Comment:
Hmm, these docstrings are now copied to both ops/models.py and here (three copies). Do we really need them repeated here in the testing harness? Or could we just point/link to the other docstring?

Reply:
Does pydoc have a good way to reference another docstring so that it gets updated naturally?

Quick googling shows things like:
https://docrep.readthedocs.io/en/latest/

For _TestingPebbleClient I probably don't care about having the docstrings much at all, since you shouldn't really have those objects at hand. I did it mostly for *me* developing it to have the immediate interface at hand.

Should we put a TODO: to do something like add a helper that just iterates the methods of the class and copies Backend.method.__doc__ to _TestingPebbleClient.method.__doc__ ?

Reply:
Yeah, a TODO sounds fine, or it's probably very straight-forward to just add the loop too. I believe Sphinx picks up dynamically-added docstrings (I tested that the other day).

Reply:
I actually just wrote _copy_docstrings which I then applied to both backends.
And 'help(ops.testing._TestingPebbleBackend)' did what I wanted. I don't know about sphinx yet, but we can check when this lands.
```
@_copy_docstrings(pebble.Client)
class _TestingPebbleClient:
```
yielded

```
class _TestingPebbleClient(builtins.object)
 |  _TestingPebbleClient(backend: ops.testing._TestingModelBackend)
 |
 |  This conforms to the interface for pebble.Client but provides canned data.
 |
 |  DO NOT use this class directly, it is used by `Harness`_ to run interactions with Pebble.
 |  `Harness`_ is responsible for maintaining the internal consistency of the values here,
 |  as the only public methods of this type are for implementing Client.
 |
 |  Methods defined here:
 |
 |  __init__(self, backend: ops.testing._TestingModelBackend)
 |      Initialize self.  See help(type(self)) for accurate signature.
 |
 |  abort_change(self, change_id: ops.pebble.ChangeID) -> ops.pebble.Change
 |      Abort change with given ID.
 |
 |  ack_warnings(self, timestamp: datetime.datetime) -> int
 |      Acknowledge warnings up to given timestamp, return number acknowledged.
...
```